File: 020_ValidParentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        tmp=[]
        dic= ['()','{}','[]']

        for i in xrange(len(s)):
            tmp.append(s[i])
            if len(tmp)>=2 and tmp[-2]+tmp[-1] in dic:
                tmp.pop()
                tmp.pop()
        if tmp == []:
            return True
        else:
            return False

if __name__ == '__main__':
    inputs ='()'
    print inputs
    k = Solution()

    r = k.isValid(inputs)
    print r

File: 020_ValidParentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {// faster
    var dict={
        "(":"0",
        "[":"0",
        "{":"0",
        ")":"(",
        "]":"[",
        "}":"{"
    };
    var i = 0;
    var tmp_s = [];
    for(i=0;i<s.length;i++){
        var tmp_char = dict[s[i]];
        if(tmp_char==="0"){
            tmp_s.push(s[i]);
        }else{
            if(dict[s[i]]!=tmp_s.pop()){
                return(false);
            };
        };
    };
    if(tmp_s.length===0){
        return(true);
    }else{
        return(false);
    };
};

var isValid_2 = function(s) {// other solution
    var i=0;
    var tmp_stack =[];
    for(i=0;i<s.length;i++){
        switch(s[i]){
            case "(":
                tmp_stack.push(s[i]);
                break;
            case "{":
                tmp_stack.push(s[i]);
                break;
            case "[":
                tmp_stack.push(s[i]);
                break;
            case ")":
                if(tmp_stack.pop()!="("){
                    return(false);
                };
                break;
            case "}":
                if(tmp_stack.pop()!="{"){
                    return(false);
                };
                break;
            case "]":
                if(tmp_stack.pop()!="["){
                    return(false);
                };
                break;
        };
    };
    if(tmp_stack.length===0){
        return(true);
    }else{
        return(false);
    };  
};

var string_example ="{[]}";
console.log(isValid(string_example));